# **Data science 2- Homework 2 solution**
***

In [2]:
# imports
from ucimlrepo import fetch_ucirepo
import pandas as pd
import numpy as np

## **1. Load the "Individual household electric power consumption" dataset** 
***

First, we will load the dataset as advised in the following link of [UCI Machine Learning Repository](https://archive.ics.uci.edu/dataset/235/individual+household+electric+power+consumption):

In [3]:
# fetch dataset 
individual_household_electric_power_consumption = fetch_ucirepo(id=235) 
  
# data (as pandas dataframes) 
X = individual_household_electric_power_consumption.data.features 
y = individual_household_electric_power_consumption.data.targets   
# variable information 
display(individual_household_electric_power_consumption.variables)

C:\Data_Science2_Lab\Homework2\homework2_venv\Lib\site-packages\ucimlrepo\fetch.py:97: DtypeWarning: Columns (2,3,4,5,6,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(data_url)


name     role         type demographic description units  \
0                   Date  Feature         Date        None        None  None   
1                   Time  Feature  Categorical        None        None  None   
2    Global_active_power  Feature   Continuous        None        None  None   
3  Global_reactive_power  Feature   Continuous        None        None  None   
4                Voltage  Feature   Continuous        None        None  None   
5       Global_intensity  Feature   Continuous        None        None  None   
6         Sub_metering_1  Feature   Continuous        None        None  None   
7         Sub_metering_2  Feature   Continuous        None        None  None   
8         Sub_metering_3  Feature   Continuous        None        None  None   

  missing_values  
0             no  
1             no  
2             no  
3             no  
4             no  
5             no  
6             no  
7             no  
8             no

We can see the variables of the dataset. This dataset contains 9 features in order to predict the individual household electric power consumption. We can see features like Date, time, active and reactive power, voltage, etc...
The measurements were taken from one household with a one-minute sampling rate over a period of almost 4 years. 
(global_active_power*1000/60 - sub_metering_1 - sub_metering_2 - sub_metering_3) represents the active energy consumed every minute (in watt/hour) in the household by electrical equipment not measured in sub-meters 1, 2 and 3.


In [4]:
X.head()

Date      Time Global_active_power Global_reactive_power  Voltage  \
0  16/12/2006  17:24:00               4.216                 0.418  234.840   
1  16/12/2006  17:25:00               5.360                 0.436  233.630   
2  16/12/2006  17:26:00               5.374                 0.498  233.290   
3  16/12/2006  17:27:00               5.388                 0.502  233.740   
4  16/12/2006  17:28:00               3.666                 0.528  235.680   

  Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0           18.400          0.000          1.000            17.0  
1           23.000          0.000          1.000            16.0  
2           23.000          0.000          2.000            17.0  
3           23.000          0.000          1.000            17.0  
4           15.800          0.000          1.000            17.0

we can see That Date has a date format of DD/MM/YYYY, a timestamp of HH:MM:SS  and 3 sub-meters results that allows the landlord to bill each tenant separately.

In [5]:
display(individual_household_electric_power_consumption.data.original)

Date      Time Global_active_power Global_reactive_power  \
0        16/12/2006  17:24:00               4.216                 0.418   
1        16/12/2006  17:25:00               5.360                 0.436   
2        16/12/2006  17:26:00               5.374                 0.498   
3        16/12/2006  17:27:00               5.388                 0.502   
4        16/12/2006  17:28:00               3.666                 0.528   
...             ...       ...                 ...                   ...   
2075254  26/11/2010  20:58:00               0.946                   0.0   
2075255  26/11/2010  20:59:00               0.944                   0.0   
2075256  26/11/2010  21:00:00               0.938                   0.0   
2075257  26/11/2010  21:01:00               0.934                   0.0   
2075258  26/11/2010  21:02:00               0.932                   0.0   

         Voltage Global_intensity Sub_metering_1 Sub_metering_2  \
0        234.840           18.400          0.000          1.000   
1        233.630           23.000          0.000          1.000   
2        233.290           23.000          0.000          2.000   
3        233.740           23.000          0.000          1.000   
4        235.680           15.800          0.000          1.000   
...          ...              ...            ...            ...   
2075254   240.43              4.0            0.0            0.0   
2075255    240.0              4.0            0.0            0.0   
2075256   239.82              3.8            0.0            0.0   
2075257    239.7              3.8            0.0            0.0   
2075258   239.55              3.8            0.0            0.0   

         Sub_metering_3  
0                  17.0  
1                  16.0  
2                  17.0  
3                  17.0  
4                  17.0  
...                 ...  
2075254             0.0  
2075255             0.0  
2075256             0.0  
2075257             0.0  
2075258             0.0  

[2075259 rows x 9 columns]

In [6]:
display(individual_household_electric_power_consumption.data.targets)

None

we can see that we have no target variables, so as described in the dataset documentation we need to calculate the target using the above formula: (global_active_power*1000/60 - sub_metering_1 - sub_metering_2 - sub_metering_3)

In [7]:
individual_household_electric_power_consumption.metadata

{'uci_id': 235,
 'name': 'Individual Household Electric Power Consumption',
 'repository_url': 'https://archive.ics.uci.edu/dataset/235/individual+household+electric+power+consumption',
 'data_url': 'https://archive.ics.uci.edu/static/public/235/data.csv',
 'abstract': 'Measurements of electric power consumption in one household with a one-minute sampling rate over a period of almost 4 years. Different electrical quantities and some sub-metering values are available.',
 'area': 'Physics and Chemistry',
 'tasks': ['Regression', 'Clustering'],
 'characteristics': ['Multivariate', 'Time-Series'],
 'num_instances': 2075259,
 'num_features': 9,
 'feature_types': ['Real'],
 'demographics': [],
 'target_col': None,
 'index_col': None,
 'has_missing_values': 'no',
 'missing_values_symbol': None,
 'year_of_dataset_creation': 2006,
 'last_updated': 'Fri Mar 08 2024',
 'dataset_doi': '10.24432/C58K54',
 'creators': ['Georges Hebrail', 'Alice Berard'],
 'intro_paper': None,
 'additional_info': {'s

In [13]:
print(f"{individual_household_electric_power_consumption.data.original.isna().sum().sum() / individual_household_electric_power_consumption.data.original.shape[0] * 100}% of the data is missing")

1.2518437457686005% of the data is missing


we can see that 1.25% of the data is missing, and we will need to think of how to handle that.